fw/rbcx-coprocessor/src/MotorController.cpp

```diff
@@ -0,0 +1,187 @@
+    ocInit.OCNPolarity = LL_TIM_OCPOLARITY_HIGH;
+    ocInit.OCIdleState = LL_TIM_OCIDLESTATE_HIGH;
+    ocInit.OCNIdleState = LL_TIM_OCIDLESTATE_HIGH;
+
+    LL_TIM_Init(pwmTimer, &pwmInit);
+    for (uint16_t channel = LL_TIM_CHANNEL_CH1; channel != 0; channel <<= 4) {
+        LL_TIM_OC_Init(pwmTimer, channel, &ocInit);
+        LL_TIM_OC_EnablePreload(pwmTimer, channel);
+    }
+
+    LL_TIM_SetOffStates(pwmTimer, LL_TIM_OSSI_DISABLE, LL_TIM_OSSR_ENABLE);
+    LL_TIM_GenerateEvent_UPDATE(pwmTimer);
+    LL_TIM_EnableAllOutputs(pwmTimer);
+    LL_TIM_EnableCounter(pwmTimer);
+
+    for (int motorIndex : { 0, 1, 2, 3 }) {
+        setMotorPower(motorIndex, 0, false);
+    }
+
+    LL_TIM_ENCODER_InitTypeDef encInit;
+    encInit.EncoderMode = LL_TIM_ENCODERMODE_X4_TI12;
+    encInit.IC1Polarity = LL_TIM_IC_POLARITY_RISING;
+    encInit.IC1ActiveInput = LL_TIM_ACTIVEINPUT_DIRECTTI;
+    encInit.IC1Prescaler = LL_TIM_ICPSC_DIV1;
+    encInit.IC1Filter = LL_TIM_IC_FILTER_FDIV1;
+    encInit.IC2Polarity = LL_TIM_IC_POLARITY_RISING;
+    encInit.IC2ActiveInput = LL_TIM_ACTIVEINPUT_DIRECTTI;
+    encInit.IC2Prescaler = LL_TIM_ICPSC_DIV1;
+    encInit.IC2Filter = LL_TIM_IC_FILTER_FDIV1;
+    for (auto timer : encoderTimer) {
+        LL_TIM_ENCODER_Init(timer, &encInit);
+        LL_TIM_EnableCounter(timer);
+    }
+
+    motorSem = xSemaphoreCreateBinaryStatic(&motorSemStruct);
+    xSemaphoreGive(motorSem);
+    motorTask.start("motors", 2, taskFunc);
+}
+
+static void taskFunc() {
+    while (true) {
+        auto now = xTaskGetTickCount();
+        if (xSemaphoreTake(motorSem, pdMS_TO_TICKS(100) == pdTRUE)) {
+            for (int m : { 0, 1, 2, 3 }) {
+                uint16_t encTicks = LL_TIM_GetCounter(encoderTimer[m]);
+                auto action = motor[m].poll(encTicks);
+                setMotorPower(m, action.first, action.second);
+            }
+            xSemaphoreGive(motorSem);
+        }
+        vTaskDelayUntil(&now, pdMS_TO_TICKS(1000 / motorLoopFreq));
+    }
+}
+
+void motorDispatch(const CoprocReq_MotorReq& request) {
+    if (request.motorIndex > 3) {
+        return;
+    }
+
+    auto& targetMotor = motor[request.motorIndex];
+
+    if (xSemaphoreTake(motorSem, pdMS_TO_TICKS(100)) == pdTRUE) {
+        switch (request.which_motorCmd) {
+        case CoprocReq_MotorReq_setPower_tag:
+            targetMotor.setTargetPower(request.motorCmd.setPower);
+            break;
+        case CoprocReq_MotorReq_setBrake_tag:
+            targetMotor.setTargetBrakingPower(request.motorCmd.setBrake);
+            break;
+        case CoprocReq_MotorReq_setVelocity_tag: {
+            auto ticksPerSec = request.motorCmd.setVelocity;
+            if (ticksPerSec > SHRT_MAX || ticksPerSec < SHRT_MIN) {
+                DEBUG("Motor %d target velocity out of range <-32768; 32767> "
+                      "(%d).\n",
+                    int(request.motorIndex), int(ticksPerSec));
+                return;
```

Comment:
return bez odemknutí mutexu/semaphore. Hodil by se nato udělat nějaký wrapper jako je `std::lock_guard`, aby to odemykalo samo.

Reply:
Udělal jsem MutexWrapper, kterej jde použít se std::scoped_lock a spol.